src/Basics/Extra.elm

```diff
@@ -277,3 +284,163 @@ This combines two arguments into a single pair.
 uncurry : (a -> b -> c) -> ( a, b ) -> c
 uncurry f ( a, b ) =
     f a b
+
+
+{-| Create an ordering function that can be used to sort
+lists by multiple dimensions, by flattening multiple ordering functions into one.
+
+This is equivalent to `ORDER BY` in SQL. The ordering function will order
+its inputs based on the order that they appear in the `List (a -> a -> Order)` argument.
+
+    type alias Pen =
+        { model : String
+        , tipWidthInMillimeters : Float
+        }
+
+    pens : List Pen
+    pens =
+        [ Pen "Pilot Hi-Tec-C Gel" 0.4
+        , Pen "Morning Glory Pro Mach" 0.38
+        , Pen "Pilot Hi-Tec-C Coleto" 0.5
+        ]
+
+    order : Pen -> Pen -> Order
+    order =
+        orderBy [ toOrder .tipWidthInMillimeters, toOrder .model ]
+
+    List.sortWith order pens
+    --> [ Pen "Morning Glory Pro Mach" 0.38
+    --> , Pen "Pilot Hi-Tec-C Gel" 0.4
+    --> , Pen "Pilot Hi-Tec-C Coleto" 0.5
+    --> ]
+
+If our `Pen` type alias above was represented a row in a database table, our `order` function as defined above would be equivalent
+to this SQL clause:
+
+    ORDER BY tipWidthInMillimeters, model
+
+-}
+orderBy : List (a -> a -> Order) -> (a -> a -> Order)
+orderBy comparators =
+    \a b ->
+        let
+            step : (a -> a -> Order) -> Order -> Order
+            step comparator acc =
+                case acc of
+                    EQ ->
+                        comparator a b
+
+                    _ ->
+                        acc
+        in
+        List.foldl step EQ comparators
```

Comment:
Something like this would be a short-circuiting version, ie. not traversing all the comparators if we've already got the LT or GT result:
```suggestion
       case comparators of
           [] ->
               EQ
               
           comparator :: rest ->
               case comparator a b of
                   EQ ->
                       orderBy rest a b
                       
                   other ->
                       other
```

Reply:
Wow, I really like this - much easier to follow.

I felt like there was a simpler solution than using a `foldl`, but I wasn't quite there. Thank you!